### PR TITLE
client: remove WasCreated method from Result

### PR DESCRIPTION
**Pull request by 0xmichalis**
@smarterclayton @krousey ptal

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35209)

### CODE REVIEW

**Comment by smarterclayton** [comment]
I prefer just allowing the status code to be accessed.  We aren't hiding
the fact that we are talking to an HTTP server.

On Thu, Oct 20, 2016 at 11:56 AM, Michail Kargakis <notifications@github.com

**Comment by krousey** [comment]
The fact that `WasCreated` even exists irks me a little, and I would entertain the idea of removing it if it doesn't break clients. And we certainly don't need a `HasStatusCode`. That need can be completely fulfilled by the `StatusCode` accessor and a comparison.

**Comment by kargakis** [comment]
I just removed WasCreated and used [StatusCode](https://github.com/kubernetes/kubernetes/blob/eeae8b59752a0420138e3a47013a9def8080fee5/pkg/client/restclient/request.go#L1080) which does what I want without the need to export the field.

**Comment by smarterclayton** [comment]
@kargakis in the next rebase you get to fix this usage in openshift

**Comment by kargakis** [comment]
> @kargakis in the next rebase you get to fix this usage in openshift

We don't use WasCreated anywhere and I am fixing our client in https://github.com/openshift/origin/pull/11500. Not sure if you want something more.

**Comment by smarterclayton** [comment]
Why does it irk you?  That it's too specific (when status code is more
general)?

Update is one fourth of the http verbs, and PUT to create is part of the
intent of the verb.  Just not common in our use because we don't really
ever do idempotent updates or have a reason to use blind update today.

On Oct 20, 2016, at 12:21 PM, krousey <notifications@github.com> wrote:

The fact that WasCreated even exists irks me a little, and I would
entertain the idea of removing it if it doesn't break clients. And we
certainly don't need a HasStatusCode. That need can be completely fulfilled
by the StatusCode accessor and a comparison.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/35209#issuecomment-255155100>,
or mute the thread

**Comment by smarterclayton** [comment]
Hrm, looks like Jordan refactored the rest of it out.

**Comment by krousey** [comment]
>Why does it irk you?  That it's too specific (when status code is more
general)?

That's pretty much it. It needlessly expands the API surface of the client without really adding any new functionality.

```go
created := false
err = c.Put().Namespace("ns").Resource("pods").Name("foo").Body(pod).Do().WasCreated(&created)
...
if !created {
   // stuff
}
```

is not a huge win over

```go
var status int
err = c.Put().Namespace("ns").Resource("pods").Name("foo").Body(pod).Do().StatusCode(&status)
...
if status != http.StatusCreated {
   // stuff
}
```

>Update is one fourth of the http verbs, and PUT to create is part of the
intent of the verb.  Just not common in our use because we don't really
ever do idempotent updates or have a reason to use blind update today.

I can't find a single use of it in our code base outside of tests. I suspect many others don't use the restclient directly, rather they probably use the generated clients or dynamic client on top of the restclient. However I can't guarantee this, which is why I added "if it doesn't break clients."

**Comment by smarterclayton** [comment]
I'm fine with removing it, just pointing out it had a purpose originally
that we ended up not doing (because blind update is less useful than PATCH)

File: pkg/client/restclient/client_test.go [hunk 1]
```diff
@@ -102,65 +102,6 @@ func TestDoRequestFailed(t *testing.T) {
 	}
 }
 
-func TestDoRequestCreated(t *testing.T) {
```
Comment (krousey) [inline pkg/client/restclient/client_test.go]: Why remove these tests? Couldn't you just change them to check the status code?

File: pkg/client/restclient/request_test.go [hunk 2]
```diff
@@ -1198,109 +1198,6 @@ func TestDoRequestNewWayObj(t *testing.T) {
...
-	if wasCreated {
-		t.Errorf("expected object was created")
-	}
-	tmpStr := string(reqBodyExpected)
-	requestURL := testapi.Default.ResourcePathWithPrefix("foo/bar/baz", "", "", "")
-	requestURL += "?timeout=1s"
-	fakeHandler.ValidateRequest(t, requestURL, "POST", &tmpStr)
-}
-
-func TestWasCreated(t *testing.T) {
```
Comment (krousey) [inline pkg/client/restclient/request_test.go]: I think this is probably the only test that should be removed.

**Review by krousey — CHANGES_REQUESTED** [review]

**Comment by kargakis** [comment]
@krousey updated ptal

File: pkg/client/restclient/request_test.go [hunk 3]
```diff
@@ -1246,7 +1246,7 @@ func TestDoRequestNewWayFile(t *testing.T) {
 	} else if !api.Semantic.DeepDerivative(expectedObj, obj) {
 		t.Errorf("Expected: %#v, got %#v", expectedObj, obj)
 	}
-	if wasCreated {
+	if statusCode == 202 {
```
Comment (krousey) [inline pkg/client/restclient/request_test.go]: minor nit: use `http.StatusCreated`

File: pkg/client/restclient/client_test.go [hunk 4]
```diff
@@ -110,8 +110,8 @@ func TestDoRequestCreated(t *testing.T) {
 	if err != nil {
 		t.Fatalf("unexpected error: %v", err)
 	}
-	created := false
-	body, err := c.Get().Prefix("test").Do().WasCreated(&created).Raw()
+	var statusCode int
+	body, err := c.Get().Prefix("test").Do().StatusCode(&statusCode).Raw()
```
Comment (krousey) [inline pkg/client/restclient/client_test.go]: Ok. Now I'm really confused. `created` and now `statusCode` don't really seem to be used/validated ANYWHERE. In `validate`, there's a check against `testParam.actualCreated`, and I can't find anywhere where that variable is set.

Shouldn't these `TestParam`s have a field initializer `actualCreate: statusCode == http.StatusCreated,` in them?

**Review by krousey — CHANGES_REQUESTED** [review]
There seem to be some tests that had broken logic.

**Comment by kargakis** [comment]
Won't get to this anytime soon since it's not a high priority.

**Pull request closed without merging**